File: CSharp/ModulusTreeCryptography/ModulusTreeCryptography/ModulusDecryption.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;

namespace ModulusTreeCryptography
{
    public static class ModulusDecryption
    {
        public static BigInteger GetValueAtIndex(ModulusTreeKey key, BigInteger treeIndex)
        {
            if (treeIndex == 0)
                return 1;

            var modulusPath = key.GetModulusPathToIndex(treeIndex).ToList();
            var value = key.GetValueByModulusPath(modulusPath);
            var level = key.Levels[modulusPath.Count - 1];
            var size = key.Primorials[level.Prime];
            while (value < level.MinValue)
            {
                value += size;
            }
            return --value;
        }

        public static BigInteger GetValueByModulusPath(this ModulusTreeKey key, IEnumerable<BigInteger> modulusPath) // modulusPath: 0, 1, 3, 4
        {
            var path = modulusPath.ToList();
            var crtValues = key.Primes
                .Zip(path, CRTValues.Create)
                .Take(path.Count)
                .ToList();

            var value = CRT.ChineseRemainder(
                crtValues.Select(v => v.Remainder).ToList(),
                crtValues.Select(v => v.Modulus).ToList());

            return value;
        }

        public static IEnumerable<BigInteger> GetModulusPathToIndex(this ModulusTreeKey key, BigInteger treeIndex)
        {
            var level = key.GetLevelContainingIndex(treeIndex);

            var groupInfos = new List<GroupInfo>();

            var currentValue = key.GetGroupInfoForIndex(treeIndex);
            groupInfos.Add(currentValue);

            var index = level.Index;
            for (int i = index; i > 0; i--)
            {
                var currentLevel = key.Levels[i - 1];
                var newValue = key.GetGroupInfoForIndexOnLevel(currentLevel, currentValue.GroupIndex);
                groupInfos.Add(newValue);
                currentValue = newValue;
            }

            var modulusPath = new List<BigInteger>();
            modulusPath.AddRange(groupInfos.Select(gi => gi.IndexInGroup));
            modulusPath.Reverse();
            return modulusPath;
        }

        private static GroupInfo GetGroupInfoForIndex(this ModulusTreeKey key, BigInteger treeIndex)
        {
            var level = key.GetLevelContainingIndex(treeIndex);
            var parentTreeSize = level.MinValue - 1; // 9
            var levelIndex = treeIndex - parentTreeSize; // 22 - 9 = 13
            return key.GetGroupInfoForIndexOnLevel(level, levelIndex);
        }

        private static GroupInfo GetGroupInfoForIndexOnLevel(this ModulusTreeKey key, Level level, BigInteger levelIndex)
        {
            BigInteger indexInGroup; // 13 % 5 = 3
            var groupIndex = BigInteger.DivRem(levelIndex, level.NodesPerGroup, out indexInGroup); // 13 / 5 = 2

            return new GroupInfo(groupIndex, indexInGroup);
        }
        
        class GroupInfo
        {
            public GroupInfo(BigInteger groupIndex, BigInteger indexInGroup)
            {
                this.GroupIndex = groupIndex;
                this.IndexInGroup = indexInGroup;
            }

            public BigInteger GroupIndex { get; private set; }
            public BigInteger IndexInGroup { get; private set; }
        }

        class CRTValues
        {
            public static CRTValues Create(BigInteger modulus, BigInteger remainder)
            {
                var crtValue = new CRTValues
                               {
                                   Modulus = modulus, 
                                   Remainder = remainder
                               };
                return crtValue;
            }

            public BigInteger Modulus { get; set; }
            public BigInteger Remainder { get; set; }
        }
    }
}

File: Java/src/edu/byui/cs480/modulustree/ModulusDecryption.java
package edu.byui.cs480.modulustree;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by clay on 5/16/14.
 */
public class ModulusDecryption
{
    public static BigInteger getValueAtIndex(ModulusTreeKey pKey, BigInteger pTreeIndex)
    {
        if (pTreeIndex.compareTo(BigInteger.ZERO) == 0)
        {
            return BigInteger.ONE;
        }
        List<BigInteger> modulusPath = getModulusPathToIndex(pKey, pTreeIndex);
        BigInteger value = getValueByModulusPath(pKey, modulusPath);
        Level level = pKey.getmLevels().get(modulusPath.size() - 1);
        BigInteger size = pKey.getmPrimorials().get(level.getPrime());
        while (value.compareTo(level.getMinValue()) == -1)
        {
            value = value.add(BigInteger.ONE);
        }
        return value.subtract(BigInteger.ONE);
    }

    public static BigInteger getValueByModulusPath(ModulusTreeKey pKey, List<BigInteger> pModulusPath)
    {
        List<BigInteger> path = pModulusPath;
        List<CRTValues> crtValues = Utilities.zip(pKey.getmPrimes().stream(),
                                                   path.stream(),
                                                   (prime, pathVal) -> new CRTValues(prime, pathVal)).
                                                   limit(path.size()).
                                                   collect(Collectors.toList());
        try
        {
            BigInteger[] value = BigIntegerMath.solveCRT(
                    crtValues.stream().map(v -> v.getRemainder()).toArray(BigInteger[]::new),
                    crtValues.stream().map(v -> v.getModulus()).toArray(BigInteger[]::new));

//            BigInteger value = CRT.ChineseRemainder(
//                    crtValues.stream().map(v -> v.getRemainder()).collect(Collectors.toList()),
//                    crtValues.stream().map(v -> v.getModulus()).collect(Collectors.toList()));
            return value[0];
        }
        catch (Exception ex)
        {
            System.out.println(ex.getMessage());
            return null;
        }
    }

    public static List<BigInteger> getModulusPathToIndex(ModulusTreeKey pKey, BigInteger pTreeIndex)
    {
        Level level = Level.getLevelContainingIndex(pKey.getmLevels(), pTreeIndex);
        List<GroupInfo> groupInfos = new ArrayList<GroupInfo>();

        GroupInfo currentValue = getGroupInfoForIndex(pKey, pTreeIndex);
        groupInfos.add(currentValue);
        int index = level.getIndex();
        for (int i = index; i > 0; i--)
        {
            Level currentLevel = pKey.getmLevels().get(i - 1);
            GroupInfo newValue = getGroupInfoForIndexOnLevel(pKey, currentLevel, currentValue.getGroupIndex());
            groupInfos.add(newValue);
            currentValue = newValue;
        }
        List<BigInteger> modulusPath = new ArrayList<BigInteger>();
        modulusPath.addAll(groupInfos.stream().
                                      map(gi -> gi.getIndexInGroup()).
                                      collect(Collectors.toList()));
        Collections.reverse(modulusPath);
        return modulusPath;
    }

    public static GroupInfo getGroupInfoForIndex(ModulusTreeKey pKey, BigInteger pTreeIndex)
    {
        Level level = Level.getLevelContainingIndex(pKey.getmLevels(), pTreeIndex);
        BigInteger parentTreeSize = level.getMinValue().subtract(BigInteger.ONE);
        BigInteger levelIndex = pTreeIndex.subtract(parentTreeSize);
        return getGroupInfoForIndexOnLevel(pKey, level, levelIndex);

    }

    public static GroupInfo getGroupInfoForIndexOnLevel(ModulusTreeKey pKey, Level pLevel, BigInteger pLevelIndex)
    {
        BigInteger indexInGroup;
        BigInteger[] groupIndex = pLevelIndex.divideAndRemainder(pLevel.getNodesPerGroup());
        return new GroupInfo(groupIndex[0], groupIndex[1]);
    }

    private static class GroupInfo
    {
        private BigInteger mGroupIndex;
        private BigInteger mIndexInGroup;

        public BigInteger getGroupIndex() {
            return mGroupIndex;
        }

        public void setGroupIndex(BigInteger pGroupIndex) {
            mGroupIndex = pGroupIndex;
        }

        public BigInteger getIndexInGroup() {
            return mIndexInGroup;
        }

        public void setIndexInGroup(BigInteger pIndexInGroup) {
            mIndexInGroup = pIndexInGroup;
        }

        public GroupInfo(BigInteger pGroupIndex, BigInteger pIndexInGroup)
        {
            mGroupIndex = pGroupIndex;
            mIndexInGroup = pIndexInGroup;
        }
    }

    private static class CRTValues
    {
        private BigInteger mModulus;
        private BigInteger mRemainder;

        public BigInteger getRemainder() {
            return mRemainder;
        }

        public void setRemainder(BigInteger pRemainder) {
            mRemainder = pRemainder;
        }

        public BigInteger getModulus() {
            return mModulus;
        }

        public void setModulus(BigInteger pModulus) {
            mModulus = pModulus;
        }



        public CRTValues(BigInteger pModulus, BigInteger pRemainder)
        {
            mModulus = pModulus;
            mRemainder = pRemainder;
        }
    }
}
